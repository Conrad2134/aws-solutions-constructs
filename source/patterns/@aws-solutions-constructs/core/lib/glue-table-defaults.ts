/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnDatabase, CfnTable } from "@aws-cdk/aws-glue";
import { Aws, Construct } from "@aws-cdk/core";

/**
 *
 * @param scope
 * @param database
 * @param fieldSchema
 * @param sourceType
 * @param parameters - Key value pairs of parmaeters. If source type is 'Kinesis', pass Kinesis Data Stream name with key
 *  as 'STREAM_NAME'. Example: {STREAM_NAME: 'KinesisStreamConstrct.streamName'}
 */
export function DefaultGlueTableProps(scope: Construct, database: CfnDatabase, fieldSchema: CfnTable.ColumnProperty [],
                                      sourceType: string, parameters?: any): CfnTable {
  let _glueTable: CfnTable;
  if (sourceType === 'Kinesis') {
    const kinesisStreamName = parameters.STREAM_NAME;
    _glueTable = new CfnTable(scope, 'GlueTable', {
      catalogId: database.catalogId,
      databaseName: database.ref,
      tableInput: {
        storageDescriptor: {
          columns: fieldSchema,
          location: kinesisStreamName,
          inputFormat: "org.apache.hadoop.mapred.TextInputFormat",
          outputFormat: "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
          compressed: false,
          numberOfBuckets: -1,
          serdeInfo: {
            serializationLibrary: "org.openx.data.jsonserde.JsonSerDe",
            parameters: {
              paths: '.'
            }
          },
          parameters: {
            endpointUrl: `https://kinesis.${Aws.REGION}.amazonaws.com`,
            streamName: kinesisStreamName,
            typeOfData: "kinesis"
          }
        },
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          classication: 'json'
        }
      }
    });
  } else {
    throw Error('Source Type not Supported. Valid Source Type not provided');
  }

  return _glueTable!;
}