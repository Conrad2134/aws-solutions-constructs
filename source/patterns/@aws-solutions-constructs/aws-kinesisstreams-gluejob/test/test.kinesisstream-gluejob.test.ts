/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { ResourcePart, SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { CfnJob } from '@aws-cdk/aws-glue';
import { Bucket } from '@aws-cdk/aws-s3';
import { Stack } from "@aws-cdk/core";
import { buildKinesisStream } from '@aws-solutions-constructs/core';
import { KinesisStreamGlueJob, KinesisStreamGlueJobProps } from '../lib';

// --------------------------------------------------------------
// Pattern minimal deployment
// --------------------------------------------------------------
test('Pattern minimal deployment', () => {
    // Initial setup
    const stack = new Stack();
    const props: KinesisStreamGlueJobProps = {
        glueJobProps: {
            command: KinesisStreamGlueJob.createGlueJobCommand(stack, 'testJob', '3'),
            role: KinesisStreamGlueJob.createGlueJobRole(stack).roleArn,
            securityConfiguration: 'testSecConfig'
        }
    };
    new KinesisStreamGlueJob(stack, 'test-kinesisstreams-lambda', props);
    // Assertion 1
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

    // check for role creation
    expect(stack).toHaveResourceLike('AWS::IAM::Role', {
        Properties: {
            AssumeRolePolicyDocument: {
                Statement: [{
                    Action: "sts:AssumeRole",
                    Effect: "Allow",
                    Principal: {
                        Service: "glue.amazonaws.com",
                    },
                }],
                Version: "2012-10-17",
            },
        Description: "Service role that Glue custom ETL jobs will assume for exeuction",
        },
        Type: "AWS::IAM::Role"
    }, ResourcePart.CompleteDefinition);

    // check for bucket creation
    expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
        Properties: {
            BucketEncryption: {
                ServerSideEncryptionConfiguration: [{
                    ServerSideEncryptionByDefault: {
                        SSEAlgorithm: "AES256",
                    },
                }],
            },
            LoggingConfiguration: {
                DestinationBucketName: {
                    Ref: "S3LoggingBucket800A2B27",
                },
            },
            PublicAccessBlockConfiguration: {
                BlockPublicAcls: true,
                BlockPublicPolicy: true,
                IgnorePublicAcls: true,
                RestrictPublicBuckets: true,
            },
            VersioningConfiguration: {
                Status: "Enabled",
            },
        },
        Type: "AWS::S3::Bucket",
        UpdateReplacePolicy: "Retain"
    }, ResourcePart.CompleteDefinition);

    // check for Kinesis Stream
    expect(stack).toHaveResourceLike('AWS::Kinesis::Stream', {
        Properties: {
            RetentionPeriodHours: 24,
            ShardCount: 1,
            StreamEncryption: {
                EncryptionType: "KMS",
                KeyId: "alias/aws/kinesis",
            },
        },
        Type: "AWS::Kinesis::Stream"
    }, ResourcePart.CompleteDefinition);

    // check policy to allow read access to Kinesis Stream
    expect(stack).toHaveResourceLike('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [{
                Action: [
                    "kinesis:DescribeStreamSummary",
                    "kinesis:GetRecords",
                    "kinesis:GetShardIterator",
                    "kinesis:ListShards",
                    "kinesis:SubscribeToShard",
                ],
                Effect: "Allow",
                Resource: {
                    "Fn::GetAtt": [
                        "testkinesisstreamslambdaKinesisStream374D6D56",
                        "Arn",
                    ],
                },
            }],
            Version: "2012-10-17"
        }
    }, ResourcePart.Properties);
});

// --------------------------------------------------------------
// Test if existing Glue Job is provided
// --------------------------------------------------------------
test('Pattern minimal deployment', () => {
    // Initial setup
    const stack = new Stack();
    const existingCfnJob = new CfnJob(stack, 'ExistingJob', {
        command: KinesisStreamGlueJob.createGlueJobCommand(stack, 'testJob', '3'),
        role: KinesisStreamGlueJob.createGlueJobRole(stack).roleArn,
        securityConfiguration: 'testSecConfig'
    });

    new KinesisStreamGlueJob(stack, 'test-kinesisstreams-lambda', {
        existingGlueJob: existingCfnJob
    });
    // Assertion 1
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

    // check for Kinesis Stream
    expect(stack).toHaveResourceLike('AWS::Kinesis::Stream', {
        Properties: {
            RetentionPeriodHours: 24,
            ShardCount: 1,
            StreamEncryption: {
                EncryptionType: "KMS",
                KeyId: "alias/aws/kinesis",
            },
        },
        Type: "AWS::Kinesis::Stream"
    }, ResourcePart.CompleteDefinition);

    // check policy to allow read access to Kinesis Stream
    expect(stack).toHaveResourceLike('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [{
                Action: [
                    "kinesis:DescribeStreamSummary",
                    "kinesis:GetRecords",
                    "kinesis:GetShardIterator",
                    "kinesis:ListShards",
                    "kinesis:SubscribeToShard",
                ],
                Effect: "Allow",
                Resource: {
                    "Fn::GetAtt": [
                        "testkinesisstreamslambdaKinesisStream374D6D56",
                        "Arn",
                    ],
                },
            }],
            Version: "2012-10-17"
        }
    }, ResourcePart.Properties);
});

// --------------------------------------------------------------
// Test if existing S3 bucket location is provided
// --------------------------------------------------------------
test('When S3 bucket location for script exists', () => {
    // Initial setup
    const stack = new Stack();
    const props: KinesisStreamGlueJobProps = {
        glueJobProps: {
            command: KinesisStreamGlueJob.createGlueJobCommand(stack, 'testJob', '3', new Bucket(stack, 'ScriptLocation')),
            role: KinesisStreamGlueJob.createGlueJobRole(stack).roleArn,
            securityConfiguration: 'testSecConfig'
        },
    };
    new KinesisStreamGlueJob(stack, 'test-kinesisstreams-lambda', props);
    // Assertion 1
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

    // Assert existing bucket
    expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
        DeletionPolicy: "Retain",
        Type: "AWS::S3::Bucket",
        UpdateReplacePolicy: "Retain",
    }, ResourcePart.CompleteDefinition);

    // Assert bucket policy on existing bucket
    expect(stack).toHaveResourceLike('AWS::S3::BucketPolicy', {
        Properties: {
            Bucket: {
                Ref: "ScriptLocation6396DB53",
            },
            PolicyDocument: {
                Statement: [{
                    Action: [ "s3:GetObject*", "s3:GetBucket*", "s3:List*" ],
                    Effect: "Allow",
                    Principal: {
                        Service: "glue.amazonaws.com",
                    },
                    Resource: [{
                        "Fn::GetAtt": [
                            "ScriptLocation6396DB53",
                            "Arn",
                        ]}, {
                            "Fn::Join": [
                                "", [{
                                    "Fn::GetAtt": [
                                        "ScriptLocation6396DB53",
                                        "Arn",
                                    ],
                                },
                                "/*",
                                ],
                            ],
                        },
                    ],
                }],
                Version: "2012-10-17",
            }
        },
        Type: "AWS::S3::BucketPolicy"
    }, ResourcePart.CompleteDefinition);
});

// --------------------------------------------------------------
// Test when the construct is supplied with an existing stream
// --------------------------------------------------------------
test('create glue job with existing kinesis stream', () => {
    const stack = new Stack();
    const kinesisStream = buildKinesisStream(stack, {});
    new KinesisStreamGlueJob(stack, 'existingStreamJob', {
        existingStreamObj: kinesisStream,
        glueJobProps: {
            command: KinesisStreamGlueJob.createGlueJobCommand(stack, 'JobCommand', '3'),
            role: KinesisStreamGlueJob.createGlueJobRole(stack).roleArn
        }
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

    // Assert if the Glue Job role has read access to the stream
    expect(stack).toHaveResourceLike('AWS::IAM::Policy', {
        Properties: {
            PolicyDocument: {
                Statement: [{
                    Action: [ "kinesis:DescribeStreamSummary", "kinesis:GetRecords", "kinesis:GetShardIterator",
                                "kinesis:ListShards", "kinesis:SubscribeToShard"
                    ],
                    Effect: "Allow",
                    Resource: {
                        "Fn::GetAtt":  [
                            "KinesisStream46752A3E",
                            "Arn",
                        ]
                    },
                }],
                Version: "2012-10-17",
            },
            PolicyName: "GlueJobRolePolicy422D4F2E",
            Roles: [{
                "Fn::Select": [
                    1, {
                        "Fn::Split": [
                        "/", {
                            "Fn::Select": [
                                5, {
                                    "Fn::Split": [
                                    ":", {
                                        "Fn::GetAtt": [
                                            "JobRole014917C6",
                                            "Arn",
                                        ],
                                    }]
                                },
                            ]
                        },
                    ],
                }],
            }],
        },
        Type: "AWS::IAM::Policy",
    }, ResourcePart.CompleteDefinition);
});