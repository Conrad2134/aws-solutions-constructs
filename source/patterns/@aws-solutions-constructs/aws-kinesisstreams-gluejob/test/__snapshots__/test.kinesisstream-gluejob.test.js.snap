// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pattern minimal deployment 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acArtifactHash2A1C2D40": Object {
      "Description": "Artifact hash for asset \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B": Object {
      "Description": "S3 bucket for asset \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE": Object {
      "Description": "S3 key for asset version \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GlueJobRolePolicy422D4F2E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testkinesisstreamslambdaKinesisStream374D6D56",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GlueJobRolePolicy422D4F2E",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "JobRole014917C6",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JobRole014917C6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Service role that Glue custom ETL jobs will assume for exeuction",
      },
      "Type": "AWS::IAM::Role",
    },
    "testkinesisstreamslambdaETLJob44B50676": Object {
      "Properties": Object {
        "Command": Object {
          "Name": "testJob",
          "PythonVersion": "3",
          "ScriptLocation": Object {
            "Fn::Join": Array [
              "",
              Array [
                "s3://",
                Object {
                  "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B",
                },
                "/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Role": Object {
          "Fn::GetAtt": Array [
            "JobRole014917C6",
            "Arn",
          ],
        },
        "SecurityConfiguration": "testSecConfig",
      },
      "Type": "AWS::Glue::Job",
    },
    "testkinesisstreamslambdaKinesisStream374D6D56": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "testkinesisstreamslambdaLogPolicy5FB58427": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testkinesisstreamslambdaLogPolicy5FB58427",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "JobRole014917C6",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`Test if existing Glue Job is provided 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acArtifactHash2A1C2D40": Object {
      "Description": "Artifact hash for asset \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B": Object {
      "Description": "S3 bucket for asset \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE": Object {
      "Description": "S3 key for asset version \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ExistingJob": Object {
      "Properties": Object {
        "Command": Object {
          "Name": "testJob",
          "PythonVersion": "3",
          "ScriptLocation": Object {
            "Fn::Join": Array [
              "",
              Array [
                "s3://",
                Object {
                  "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B",
                },
                "/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Role": Object {
          "Fn::GetAtt": Array [
            "JobRole014917C6",
            "Arn",
          ],
        },
        "SecurityConfiguration": "testSecConfig",
      },
      "Type": "AWS::Glue::Job",
    },
    "GlueJobRolePolicy422D4F2E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testkinesisstreamslambdaKinesisStream374D6D56",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GlueJobRolePolicy422D4F2E",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "JobRole014917C6",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JobRole014917C6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Service role that Glue custom ETL jobs will assume for exeuction",
      },
      "Type": "AWS::IAM::Role",
    },
    "testkinesisstreamslambdaKinesisStream374D6D56": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;

exports[`When S3 bucket location for script exists 1`] = `
Object {
  "Resources": Object {
    "GlueJobRolePolicy422D4F2E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testkinesisstreamslambdaKinesisStream374D6D56",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GlueJobRolePolicy422D4F2E",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "JobRole014917C6",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JobRole014917C6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Service role that Glue custom ETL jobs will assume for exeuction",
      },
      "Type": "AWS::IAM::Role",
    },
    "ScriptLocation6396DB53": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testkinesisstreamslambdaETLJob44B50676": Object {
      "Properties": Object {
        "Command": Object {
          "Name": "testJob",
          "PythonVersion": "3",
          "ScriptLocation": Object {
            "Fn::Join": Array [
              "",
              Array [
                "s3://",
                Object {
                  "Ref": "ScriptLocation6396DB53",
                },
                "/dummyfile.py",
              ],
            ],
          },
        },
        "Role": Object {
          "Fn::GetAtt": Array [
            "JobRole014917C6",
            "Arn",
          ],
        },
        "SecurityConfiguration": "testSecConfig",
      },
      "Type": "AWS::Glue::Job",
    },
    "testkinesisstreamslambdaKinesisStream374D6D56": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "testkinesisstreamslambdaLogPolicy5FB58427": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testkinesisstreamslambdaLogPolicy5FB58427",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "JobRole014917C6",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`create glue job with existing kinesis stream 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acArtifactHash2A1C2D40": Object {
      "Description": "Artifact hash for asset \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B": Object {
      "Description": "S3 bucket for asset \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE": Object {
      "Description": "S3 key for asset version \\"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GlueJobRolePolicy422D4F2E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "KinesisStream46752A3E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GlueJobRolePolicy422D4F2E",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "JobRole014917C6",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JobRole014917C6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Service role that Glue custom ETL jobs will assume for exeuction",
      },
      "Type": "AWS::IAM::Role",
    },
    "KinesisStream46752A3E": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "existingStreamJobETLJob50478D3D": Object {
      "Properties": Object {
        "Command": Object {
          "Name": "JobCommand",
          "PythonVersion": "3",
          "ScriptLocation": Object {
            "Fn::Join": Array [
              "",
              Array [
                "s3://",
                Object {
                  "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B",
                },
                "/",
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Role": Object {
          "Fn::GetAtt": Array [
            "JobRole014917C6",
            "Arn",
          ],
        },
        "SecurityConfiguration": "ETLJobSecurityConfig",
      },
      "Type": "AWS::Glue::Job",
    },
    "existingStreamJobGlueSecurityConfig88E060CD": Object {
      "Properties": Object {
        "EncryptionConfiguration": Object {
          "CloudWatchEncryption": Object {
            "CloudWatchEncryptionMode": "SSE-KMS",
            "KmsKeyArn": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":kms:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":alias/aws/glue",
                ],
              ],
            },
          },
          "JobBookmarksEncryption": Object {
            "JobBookmarksEncryptionMode": "CSE-KMS",
            "KmsKeyArn": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":kms:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":alias/aws/glue",
                ],
              ],
            },
          },
          "S3Encryptions": Array [
            Object {
              "S3EncryptionMode": "SSE-S3",
            },
          ],
        },
        "Name": "ETLJobSecurityConfig",
      },
      "Type": "AWS::Glue::SecurityConfiguration",
    },
    "existingStreamJobLogPolicy08F24C09": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "existingStreamJobLogPolicy08F24C09",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "JobRole014917C6",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
